demos/common/processing/async_pipeline.cpp

```diff
@@ -0,0 +1,217 @@
+    model->prepareInputsOutputs(cnnNetwork);
+
+    // --------------------------- 4. Loading model to the device ------------------------------------------
+    slog::info << "Loading model to the device" << slog::endl;
+    execNetwork = ie->LoadNetwork(cnnNetwork, cnnConfig.devices, cnnConfig.execNetworkConfig);
+
+    // --------------------------- 5. Create infer requests ------------------------------------------------
+    requestsPool.reset(new RequestsPool(execNetwork, cnnConfig.maxAsyncRequests));
+
+    // --------------------------- 6. Call onLoadCompleted to complete initialization of model -------------
+    model->onLoadCompleted(&execNetwork, requestsPool.get());
+}
+
+PipelineBase::~PipelineBase(){
+    waitForTotalCompletion();
+}
+
+
+void PipelineBase::waitForData(bool shouldKeepOrder){
+    std::unique_lock<std::mutex> lock(mtx);
+
+    condVar.wait(lock, [&] {return callbackException != nullptr ||
+        requestsPool->isIdleRequestAvailable() ||
+        (shouldKeepOrder ?
+            completedInferenceResults.find(outputFrameId) != completedInferenceResults.end() :
+            !completedInferenceResults.empty());
+    });
+
+    if (callbackException)
+        std::rethrow_exception(callbackException);
+}
+
+int64_t PipelineBase::submitRequest(InferenceEngine::InferRequest::Ptr request, std::shared_ptr<MetaData> metaData){
+    perfInfo.numRequestsInUse = (uint32_t)requestsPool->getInUseRequestsCount();
+
+    auto frameStartTime = std::chrono::steady_clock::now();
+
+    if (!perfInfo.startTime.time_since_epoch().count())
+    {
+        perfInfo.startTime = frameStartTime;
+    }
+
+    auto frameID = inputFrameId;
+
+    request->SetCompletionCallback([this,
+        frameStartTime,
+        frameID,
+        request,
+        metaData] {
+            {
+                std::lock_guard<std::mutex> lock(mtx);
+
+                try {
+                    InferenceResult result;
+
+                    result.frameId = frameID;
+                    result.metaData = std::move(metaData);
+                    for (std::string outName : model->getOutputsNames())
+                        result.outputsData.emplace(outName, std::make_shared<TBlob<float>>(*as<TBlob<float>>(request->GetBlob(outName))));
+                    result.startTime = frameStartTime;
+                    perfInfo.lastInferenceLatency = std::chrono::steady_clock::now() - frameStartTime;
+                    completedInferenceResults.emplace(frameID, result);
+
+                    this->requestsPool->setRequestIdle(request);
+
+                    this->onProcessingCompleted(request);
+                }
+                catch (...) {
+                    if (!this->callbackException) {
+                        this->callbackException = std::move(std::current_exception());
+                    }
+                }
+            }
+            condVar.notify_one();
+    });
+
+    inputFrameId++;
+    if (inputFrameId < 0)
```

Comment:
If the overflow happens the demo will break frame sequence anyway. But you can make `inputFrameId` unsigned to delay that.
To identify empty result you can use max uint64_t.

Reply:
frame will not break sequence (i.e. sequence will restart from zero, but processing will still work properly), check the inputFrameId and outputFrameId incrementing code. Your suggesting will not work as in theory we may reach max uint64_t value and that value will be treated as empty and will break processing.

Reply:
BTW, even if we use signed int64_t, to reach overflow in 120FPS video, we'll have to wait for 609 315 018 years. Not sure if our software will be in use by that time...

Reply:
Ok.